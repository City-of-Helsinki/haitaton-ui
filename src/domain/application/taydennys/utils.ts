import { TaydennyspyyntoFieldKey, TaydennyspyyntoField } from './types';

const order = {
  [TaydennyspyyntoFieldKey.IDENTIFICATION_NUMBER]: 1,
  [TaydennyspyyntoFieldKey.WORK_DESCRIPTION]: 2,
  [TaydennyspyyntoFieldKey.APPLICATION_KIND]: 3,
  [TaydennyspyyntoFieldKey.CLIENT_APPLICATION_KIND]: 4,
  [TaydennyspyyntoFieldKey.POSTAL_ADDRESS]: 5,
  [TaydennyspyyntoFieldKey.GEOMETRY]: 6,
  [TaydennyspyyntoFieldKey.AREA]: 7,
  [TaydennyspyyntoFieldKey.START_TIME]: 8,
  [TaydennyspyyntoFieldKey.END_TIME]: 9,
  [TaydennyspyyntoFieldKey.CUSTOMER]: 10,
  [TaydennyspyyntoFieldKey.CONTRACTOR]: 11,
  [TaydennyspyyntoFieldKey.PROPERTY_DEVELOPER]: 12,
  [TaydennyspyyntoFieldKey.REPRESENTATIVE]: 13,
  [TaydennyspyyntoFieldKey.INVOICING_CUSTOMER]: 14,
  [TaydennyspyyntoFieldKey.ATTACHMENT]: 15,
  [TaydennyspyyntoFieldKey.PROPERTY_IDENTIFICATION_NUMBER]: 16,
  [TaydennyspyyntoFieldKey.OTHER]: 17,
};

export function sortTaydennyspyyntoFields(a: TaydennyspyyntoField, b: TaydennyspyyntoField) {
  return order[a.key] - order[b.key];
}
