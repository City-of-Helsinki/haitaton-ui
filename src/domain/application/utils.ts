import { TFunction } from 'i18next';
import api from '../api/api';
import {
  AlluStatus,
  AlluStatusStrings,
  Application,
  ApplicationDeletionResult,
} from './types/application';

/**
 * Save application to Haitaton backend
 */
export async function saveApplication(data: Application) {
  // If there is no hankeTunnus when creating new cable report application
  // call /hakemukset/luo-hanke endpoint, so that hanke is generated in the backend
  const postUrl: string =
    data.hankeTunnus === null && data.applicationType === 'CABLE_REPORT'
      ? '/hakemukset/johtoselvitys'
      : '/hakemukset';

  const response = data.id
    ? await api.put<Application>(`/hakemukset/${data.id}`, data)
    : await api.post<Application>(postUrl, data);

  return response.data;
}

/**
 * Send application to Allu
 */
export async function sendApplication(applicationId: number) {
  const response = await api.post<Application>(`/hakemukset/${applicationId}/send-application`, {});
  return response.data;
}

/**
 * Check if application is sent to Allu
 */
export function isApplicationSent(alluStatus: AlluStatusStrings | null): boolean {
  return alluStatus !== null;
}

/**
 * Check if application is pending or in draft state, meaning it's not
 * yet being handled in Allu
 */
export function isApplicationPending(alluStatus: AlluStatusStrings | null): boolean {
  return (
    alluStatus === null ||
    alluStatus === AlluStatus.PENDING_CLIENT ||
    alluStatus === AlluStatus.PENDING
  );
}

export function isApplicationDraft(alluStatus: AlluStatus | null) {
  return alluStatus === null;
}

export async function cancelApplication(applicationId: number | null) {
  if (applicationId === null) return null;

  const response = await api.delete<ApplicationDeletionResult>(`/hakemukset/${applicationId}`);
  return response.data;
}

export function getAreaDefaultName(t: TFunction, index: number | undefined, areasLength: number) {
  const label = t('hakemus:labels:workArea');
  if (areasLength > 1 && index !== undefined) return `${label} ${index + 1}`;
  return label;
}
