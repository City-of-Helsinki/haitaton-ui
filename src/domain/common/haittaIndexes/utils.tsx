import { HaittaIndexData } from './types';

// Check if any of the haitta indexes have increased
export function haveHaittaIndexesIncreased(
  data: HaittaIndexData,
  previousData?: HaittaIndexData | null,
) {
  if (!previousData) {
    return false;
  }
  return (
    data.autoliikenne.indeksi > previousData.autoliikenne.indeksi ||
    data.linjaautoliikenneindeksi > previousData.linjaautoliikenneindeksi ||
    data.raitioliikenneindeksi > previousData.raitioliikenneindeksi ||
    data.pyoraliikenneindeksi > previousData.pyoraliikenneindeksi
  );
}
