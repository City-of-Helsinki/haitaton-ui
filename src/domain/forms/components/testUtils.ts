import { fireEvent, screen } from '@testing-library/dom';

export function fillNewContactPersonForm(
  options: {
    etunimi?: string;
    sukunimi?: string;
    sahkoposti?: string;
    puhelinnumero?: string;
  } = {},
) {
  const {
    etunimi = 'Matti',
    sukunimi = 'Meikäläinen',
    sahkoposti = 'matti.meikalainen@test.com',
    puhelinnumero = '0401234567',
  } = options;
  fireEvent.change(screen.getByLabelText(/etunimi/i), {
    target: { value: etunimi },
  });
  fireEvent.change(screen.getByLabelText(/sukunimi/i), {
    target: { value: sukunimi },
  });
  fireEvent.change(screen.getAllByRole('textbox', { name: /sähköposti/i })[1], {
    target: { value: sahkoposti },
  });
  fireEvent.change(screen.getAllByRole('textbox', { name: /puhelin/i })[1], {
    target: { value: puhelinnumero },
  });
}
