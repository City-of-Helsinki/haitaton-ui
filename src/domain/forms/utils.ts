import { get } from 'lodash';
import { FieldErrors, FieldPath, FieldValues, UseFormTrigger } from 'react-hook-form';
import { ObjectSchema } from 'yup';

/**
 * Change form step if validation is successful
 * @param handleStepChange function that is called if form is valid
 * @param fieldsToValidate array of form fields that are validated
 * @param trigger react-hook-form trigger function
 * @param errors form validation errors
 * @param errorsToIgnore error keys to ignore
 */
export function changeFormStep<T extends FieldValues>(
  handleStepChange: () => void,
  fieldsToValidate: FieldPath<T>[],
  trigger: UseFormTrigger<T>,
  errors?: FieldErrors<T>,
  errorsToIgnore?: string[],
) {
  // If there are form validation errors and error keys to ignore are given,
  // don't trigger validation if all errors are those to ignore.
  const ignoreErrors =
    !errors || !errorsToIgnore
      ? false
      : fieldsToValidate.every((field) => {
          const message = get(errors, field)?.message as unknown as { key: string } | undefined;
          const errorKey = message?.key;
          if (!message) {
            return true;
          }
          if (!errorKey) {
            return false;
          }
          return errorsToIgnore.includes(errorKey);
        });

  if (ignoreErrors) {
    handleStepChange();
  } else {
    trigger(fieldsToValidate, { shouldFocus: true }).then((isValid) => {
      if (isValid) {
        handleStepChange();
      }
    });
  }
}

export function isPageValid<T extends FieldValues, T2 = T>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  schema: ObjectSchema<any>,
  pageFieldPaths: FieldPath<T>[],
  formValues: T2,
): boolean {
  let isValid = true;
  for (let i = 0; i < pageFieldPaths.length; i += 1) {
    const path = pageFieldPaths[i];
    try {
      schema.validateSyncAt(path, formValues);
    } catch (error) {
      isValid = false;
      break;
    }
  }
  return isValid;
}

export function getFieldPaths<T extends FieldValues>(
  node: object | Array<object> | null,
  pathToNode: FieldPath<T>,
): FieldPath<T>[] {
  if (node === null) {
    return [];
  }

  let fieldPaths: string[] = [];
  if (Array.isArray(node)) {
    node.forEach((obj, index) => {
      fieldPaths = fieldPaths.concat(
        Object.keys(obj).map((key) => `${pathToNode}.${index}.${key}`),
      );
    });
  } else {
    fieldPaths = Object.keys(node).map((key) => `${pathToNode}.${key}`);
  }

  return fieldPaths as FieldPath<T>[];
}
