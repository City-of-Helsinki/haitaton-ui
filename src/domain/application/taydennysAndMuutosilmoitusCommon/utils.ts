import { Muutosilmoitus } from '../muutosilmoitus/types';
import { Taydennys } from '../taydennys/types';
import { JohtoselvitysData, KaivuilmoitusData } from '../types/application';
import { modifyKaivuilmoitusDataAfterReceive } from '../utils';

export function modifyHakemusAfterReceive(
  data:
    | Taydennys<JohtoselvitysData | KaivuilmoitusData>
    | Muutosilmoitus<JohtoselvitysData | KaivuilmoitusData>,
) {
  if (data.applicationData.applicationType === 'CABLE_REPORT') {
    return data;
  }
  const kaivuilmoitusData = modifyKaivuilmoitusDataAfterReceive(
    data.applicationData as KaivuilmoitusData,
  );
  return {
    ...data,
    applicationData: kaivuilmoitusData,
  };
}
