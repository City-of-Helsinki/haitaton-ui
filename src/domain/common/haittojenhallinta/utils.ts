import { HAITTA_INDEX_TYPE, HaittaIndexData } from '../haittaIndexes/types';
import { HAITTOJENHALLINTATYYPPI } from './types';

function getHaittaIndexValue(v: number | { indeksi: number }): number {
  return typeof v === 'number' ? v : v.indeksi;
}

/**
 * Sorts HAITTOJENHALLINTATYYPPI based on given tormaystarkasteluTulos.
 * For example, if tormaystarkasteluTulos has {autoliikenneindeksi: 1.0, pyoraliiikenneindeksi: 1.3, linjaautoliikenneindeksi: 0.0, raitioliikenneindeksi: 0.0},
 * the result will be [[PYORALIIKENNE, 1.3], [AUTOLIIKENNE, 1.0], [LINJAAUTOLIIKENNE, 0.0], [RAITIOLIIKENNE, 0.0]].
 */
export function sortedLiikenneHaittojenhallintatyyppi(
  tormaystarkasteluTulos: HaittaIndexData | undefined,
): [HAITTOJENHALLINTATYYPPI, number][] {
  const defaultOrder = Object.values(HAITTOJENHALLINTATYYPPI).filter(
    (type) => type !== HAITTOJENHALLINTATYYPPI.YLEINEN && type !== HAITTOJENHALLINTATYYPPI.MUUT,
  );
  if (!tormaystarkasteluTulos) {
    return defaultOrder.map((type) => [type, 0] as [HAITTOJENHALLINTATYYPPI, number]);
  }

  function keyOfIndexType(key: HAITTA_INDEX_TYPE): keyof HaittaIndexData {
    if (key === HAITTA_INDEX_TYPE.AUTOLIIKENNEINDEKSI) {
      return 'autoliikenne';
    }
    return key.toLowerCase() as keyof HaittaIndexData;
  }

  const sortedIndices = Object.values(HAITTA_INDEX_TYPE)
    .map((key) => ({
      type: key.toUpperCase().replace('INDEKSI', '') as HAITTOJENHALLINTATYYPPI,
      value: getHaittaIndexValue(tormaystarkasteluTulos[keyOfIndexType(key)]),
    }))
    .sort((a, b): number => {
      const diff = b.value - a.value;
      if (diff === 0) {
        return defaultOrder.indexOf(a.type) - defaultOrder.indexOf(b.type);
      }
      return diff;
    });

  return sortedIndices.map((item) => [item.type, item.value] as [HAITTOJENHALLINTATYYPPI, number]);
}

export function mapNuisanceEnumIndexToNuisanceIndex(index: number): number {
  if (index === 2) return 3;
  if (index === 3) return 5;
  return index;
}

/**
 * Get haitta index value for given haittojenhallinta tyyppi from tormaystarkasteluTulos
 */
export function getHaittaIndexForTyyppi(
  tormaystarkasteluTulos: HaittaIndexData | null | undefined,
  haittojenhallintaTyyppi: HAITTOJENHALLINTATYYPPI,
) {
  if (!tormaystarkasteluTulos) {
    return 0;
  }

  function getKeyOfHaittaIndexTyyppi(tyyppi: HAITTOJENHALLINTATYYPPI): keyof HaittaIndexData {
    if (tyyppi === HAITTOJENHALLINTATYYPPI.AUTOLIIKENNE) {
      return 'autoliikenne';
    }
    return `${tyyppi.toLowerCase()}indeksi` as keyof HaittaIndexData;
  }

  return getHaittaIndexValue(
    tormaystarkasteluTulos[getKeyOfHaittaIndexTyyppi(haittojenhallintaTyyppi)],
  );
}
