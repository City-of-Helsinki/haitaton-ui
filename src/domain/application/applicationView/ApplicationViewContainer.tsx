import { useNavigate } from 'react-router-dom';
import { AxiosError } from 'axios';
import { Flex } from '@chakra-ui/react';
import { Notification } from 'hds-react';
import { useTranslation } from 'react-i18next';
import ApplicationView from './ApplicationView';
import ErrorLoadingText from '../../../common/components/errorLoadingText/ErrorLoadingText';
import useHanke from '../../hanke/hooks/useHanke';
import { useApplication } from '../hooks/useApplication';
import useLinkPath from '../../../common/hooks/useLinkPath';
import { HAKEMUS_ROUTES, HAKEMUS_TAYDENNYS_ROUTES, ROUTES } from '../../../common/types/route';
import { usePermissionsForHanke } from '../../hanke/hankeUsers/hooks/useUserRightsForHanke';
import LoadingSpinner from '../../../common/components/spinner/LoadingSpinner';
import useCreateTaydennys from '../taydennys/hooks/useCreateTaydennys';
import useCreateMuutosilmoitus from '../muutosilmoitus/hooks/useCreateMuutosilmoitus';
import { MuutosLabelProvider } from '../taydennysAndMuutosilmoitusCommon/MuutosLabelContext';

type Props = {
  id: number;
};

function ApplicationViewContainer({ id }: Readonly<Props>) {
  const { t } = useTranslation();
  const { data: application, isLoading, isError, error } = useApplication(id);
  const applicationId = application?.id;
  const { data: hanke } = useHanke(application?.hankeTunnus);
  const { data: signedInUser } = usePermissionsForHanke(application?.hankeTunnus ?? undefined);
  const navigate = useNavigate();
  const getEditApplicationPath = useLinkPath(
    HAKEMUS_ROUTES[application?.applicationType ?? 'CABLE_REPORT'],
  );
  const getEditTaydennysPath = useLinkPath(
    HAKEMUS_TAYDENNYS_ROUTES[application?.applicationType ?? 'CABLE_REPORT'],
  );
  const getEditMuutosilmoitusPath = useLinkPath(ROUTES.EDIT_KAIVUILMOITUSMUUTOSILMOITUS);
  const createTaydennysMutation = useCreateTaydennys();
  const createMuutosilmoitusMutation = useCreateMuutosilmoitus();

  function editApplication() {
    if (applicationId) {
      navigate(getEditApplicationPath({ id: applicationId.toString() }));
    }
  }

  function editTaydennys() {
    if (applicationId) {
      if (!application?.taydennys) {
        // If there is no taydennys, create one
        createTaydennysMutation.mutate(applicationId, {
          onSuccess() {
            navigate(getEditTaydennysPath({ id: applicationId.toString() }));
          },
        });
      } else {
        navigate(getEditTaydennysPath({ id: applicationId.toString() }));
      }
    }
  }

  function editMuutosilmoitus() {
    if (applicationId) {
      if (!application?.muutosilmoitus) {
        // If there is no muutosilmoitus, create one
        createMuutosilmoitusMutation.mutate(applicationId, {
          onSuccess() {
            navigate(getEditMuutosilmoitusPath({ id: applicationId.toString() }));
          },
        });
      } else {
        navigate(getEditMuutosilmoitusPath({ id: applicationId.toString() }));
      }
    }
  }

  if (isLoading) {
    return (
      <Flex justify="center" mt="var(--spacing-xl)">
        <LoadingSpinner />
      </Flex>
    );
  }

  if ((isError && (error as AxiosError).response?.status === 404) || Number.isNaN(id)) {
    return <ErrorLoadingText>{t('hakemus:errors:notFound')}</ErrorLoadingText>;
  }

  if (isError) {
    return <ErrorLoadingText />;
  }

  if (!application) {
    return null;
  }

  return (
    <MuutosLabelProvider
      value={
        application.taydennys ? t('taydennys:labels:taydennys') : t('muutosilmoitus:labels:muutos')
      }
    >
      <ApplicationView
        application={application}
        hanke={hanke}
        signedInUser={signedInUser}
        onEditApplication={editApplication}
        onEditTaydennys={editTaydennys}
        creatingTaydennys={createTaydennysMutation.isLoading}
        onEditMuutosilmoitus={editMuutosilmoitus}
        creatingMuutosilmoitus={createMuutosilmoitusMutation.isLoading}
      />
      {createTaydennysMutation.isError && (
        <Notification
          position="top-right"
          dismissible
          type="error"
          closeButtonLabelText={t('common:components:notification:closeButtonLabelText')}
          onClose={() => createTaydennysMutation.reset()}
        >
          {t('common:error')}
        </Notification>
      )}
      {createMuutosilmoitusMutation.isError && (
        <Notification
          position="top-right"
          dismissible
          type="error"
          closeButtonLabelText={t('common:components:notification:closeButtonLabelText')}
          onClose={() => createMuutosilmoitusMutation.reset()}
        >
          {t('common:error')}
        </Notification>
      )}
    </MuutosLabelProvider>
  );
}

export default ApplicationViewContainer;
