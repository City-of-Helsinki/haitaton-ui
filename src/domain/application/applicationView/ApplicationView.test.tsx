import { rest } from 'msw';
import { render, screen, waitFor } from '../../../testUtils/render';
import ApplicationViewContainer from './ApplicationViewContainer';
import { waitForLoadingToFinish } from '../../../testUtils/helperFunctions';
import { server } from '../../mocks/test-server';
import { SignedInUser } from '../../hanke/hankeUsers/hankeUser';
import * as applicationApi from '../utils';
import hakemukset from '../../mocks/data/hakemukset-data';
import { cloneDeep } from 'lodash';

jest.setTimeout(60000);

test('Correct information about application should be displayed', async () => {
  render(<ApplicationViewContainer id={4} />);
  await waitForLoadingToFinish();

  expect(screen.getAllByText('Mannerheimintien kairaukset').length).toBe(2);
  expect(screen.queryByText('JS2300003')).toBeInTheDocument();
  expect(screen.queryByText('Odottaa käsittelyä')).toBeInTheDocument();
  expect(screen.queryByText('Kaikki oikeudet')).toBeInTheDocument();
});

test('Link back to related hanke should work', async () => {
  const { user } = render(<ApplicationViewContainer id={4} />);
  await waitForLoadingToFinish();

  await user.click(screen.getByRole('link', { name: 'Mannerheimintien kaukolämpö (HAI22-3)' }));

  expect(window.location.pathname).toBe('/fi/hankesalkku/HAI22-3');
});

test('Should show error notification if application is not found', async () => {
  server.use(
    rest.get('/api/hakemukset/:id', async (_, res, ctx) => {
      return res(ctx.status(404), ctx.json({ errorMessage: 'Failed for testing purposes' }));
    }),
  );

  render(<ApplicationViewContainer id={4} />);
  await waitForLoadingToFinish();

  expect(await screen.findByText('Hakemusta ei löytynyt')).toBeInTheDocument();
});

test('Should show error notification if loading application fails', async () => {
  server.use(
    rest.get('/api/hakemukset/:id', async (_, res, ctx) => {
      return res(ctx.status(500), ctx.json({ errorMessage: 'Failed for testing purposes' }));
    }),
  );

  render(<ApplicationViewContainer id={4} />);
  await waitForLoadingToFinish();

  expect(await screen.findByText('Virhe tietojen lataamisessa.')).toBeInTheDocument();
  expect(await screen.findByText('Yritä hetken päästä uudelleen.')).toBeInTheDocument();
});

test('Should be able to go editing application when editing is possible', async () => {
  const { user } = render(<ApplicationViewContainer id={1} />);

  await waitFor(() => screen.findByRole('button', { name: 'Muokkaa hakemusta' }), {
    timeout: 10000,
  });
  await user.click(screen.getByRole('button', { name: 'Muokkaa hakemusta' }));

  expect(window.location.pathname).toBe('/fi/johtoselvityshakemus/1/muokkaa');
});

test('Application edit button should not be displayed when editing is not possible', async () => {
  render(<ApplicationViewContainer id={2} />);
  await waitForLoadingToFinish();

  expect(screen.queryByRole('button', { name: 'Muokkaa hakemusta' })).not.toBeInTheDocument();
});

test('Should be able to cancel application if it is possible', async () => {
  const { user } = render(<ApplicationViewContainer id={4} />);
  await waitForLoadingToFinish();

  await waitFor(() => screen.findByRole('button', { name: 'Peru hakemus' }), { timeout: 10000 });
  await user.click(screen.getByRole('button', { name: 'Peru hakemus' }));
  await user.click(screen.getByRole('button', { name: 'Vahvista' }));

  expect(window.location.pathname).toBe('/fi/hankesalkku/HAI22-3');
  await screen.findByText('Hakemus peruttiin onnistuneesti');
  expect(screen.queryByText('Hakemus peruttiin onnistuneesti')).toBeInTheDocument();
});

test('Should not be able to cancel application if it has moved to handling in Allu', async () => {
  render(<ApplicationViewContainer id={3} />);
  await waitForLoadingToFinish();

  expect(screen.queryByRole('button', { name: 'Peru hakemus' })).not.toBeInTheDocument();
});

test('Should be able to send application if it is not already sent', async () => {
  const hakemus = cloneDeep(hakemukset[0]);
  server.use(
    rest.get(`/api/hakemukset/:id`, async (_, res, ctx) => {
      return res(ctx.status(200), ctx.json(hakemus));
    }),
    rest.post(`/api/hakemukset/:id/laheta`, async (_, res, ctx) => {
      hakemus.alluStatus = 'PENDING';
      return res(ctx.status(200), ctx.json(hakemus));
    }),
  );

  const { user } = render(<ApplicationViewContainer id={1} />);
  await waitForLoadingToFinish();

  await screen.findByRole('button', { name: 'Lähetä hakemus' });
  const sendButton = screen.getByRole('button', { name: 'Lähetä hakemus' });
  expect(sendButton).toBeEnabled();
  await user.click(sendButton);

  await screen.findByText('Hakemus lähetetty');
  expect(screen.getByText('Hakemus lähetetty')).toBeInTheDocument();
  expect(screen.queryByRole('button', { name: 'Lähetä hakemus' })).not.toBeInTheDocument();
});

test('Should not be able to send application if it has moved to handling in Allu', async () => {
  render(<ApplicationViewContainer id={3} />);
  await waitForLoadingToFinish();

  expect(screen.queryByRole('button', { name: 'Lähetä hakemus' })).not.toBeInTheDocument();
});

test('Should disable Send button if user is not a contact person on application', async () => {
  server.use(
    rest.get('/api/hankkeet/:hankeTunnus/whoami', async (_, res, ctx) => {
      return res(
        ctx.status(200),
        ctx.json<SignedInUser>({
          hankeKayttajaId: '3fa85f64-5717-4562-b3fc-2c963f66afb4',
          kayttooikeustaso: 'HAKEMUSASIOINTI',
          kayttooikeudet: ['EDIT_APPLICATIONS'],
        }),
      );
    }),
  );

  render(<ApplicationViewContainer id={1} />);

  const sendButton = await screen.findByRole('button', { name: 'Lähetä hakemus' });
  expect(sendButton).toBeDisabled();
});

test('Should not show Edit, Cancel or Send buttons if user does not have correct permission', async () => {
  server.use(
    rest.get('/api/hankkeet/:hankeTunnus/whoami', async (_, res, ctx) => {
      return res(
        ctx.status(200),
        ctx.json<SignedInUser>({
          hankeKayttajaId: '3fa85f64-5717-4562-b3fc-2c963f66afa6',
          kayttooikeustaso: 'KATSELUOIKEUS',
          kayttooikeudet: ['VIEW'],
        }),
      );
    }),
  );

  render(<ApplicationViewContainer id={1} />);

  await waitForLoadingToFinish();

  expect(screen.queryByRole('button', { name: 'Muokkaa hakemusta' })).not.toBeInTheDocument();
  expect(screen.queryByRole('button', { name: 'Peru hakemus' })).not.toBeInTheDocument();
  expect(screen.queryByRole('button', { name: 'Lähetä hakemus' })).not.toBeInTheDocument();
});

test('Should not send multiple requests if clicking application cancel confirm button many times', async () => {
  server.use(
    rest.delete('/api/hakemukset/:id', async (_, res, ctx) => {
      return res(ctx.delay(200), ctx.status(200));
    }),
  );

  const cancelApplication = jest.spyOn(applicationApi, 'cancelApplication');
  const { user } = render(<ApplicationViewContainer id={1} />);

  await waitForLoadingToFinish();
  await screen.findByRole('button', { name: 'Peru hakemus' });

  await user.click(screen.getByRole('button', { name: 'Peru hakemus' }));
  const confirmCancelButton = screen.getByRole('button', { name: 'Vahvista' });
  await user.click(confirmCancelButton);
  await user.click(confirmCancelButton);
  await screen.findByText('Hakemus peruttiin onnistuneesti');

  expect(cancelApplication).toHaveBeenCalledTimes(1);

  cancelApplication.mockRestore();
});

test('Should not send multiple requests if clicking Send button many times', async () => {
  server.use(
    rest.post('/api/hakemukset/:id/laheta', async (_, res, ctx) => {
      return res(ctx.delay(200), ctx.status(200));
    }),
  );
  const sendApplication = jest.spyOn(applicationApi, 'sendApplication');
  const { user } = render(<ApplicationViewContainer id={1} />);
  await waitForLoadingToFinish();

  const sendButton = await screen.findByRole('button', { name: 'Lähetä hakemus' });
  await user.click(sendButton);
  await user.click(sendButton);
  await screen.findByText('Hakemus on lähetetty käsiteltäväksi.');

  expect(sendApplication).toHaveBeenCalledTimes(1);

  sendApplication.mockRestore();
});
