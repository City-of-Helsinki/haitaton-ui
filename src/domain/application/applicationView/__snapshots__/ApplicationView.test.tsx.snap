// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Excavation notification application view Shows correct information in sidebar 1`] = `
<div
  class="sideBar"
  data-testid="application-view-sidebar"
>
  <div
    class="css-11yjvs7"
  >
    <header
      class="mapHeader"
    >
      <div
        class="mapHeader__inner"
      >
        <svg
          aria-hidden="true"
          aria-label="location"
          class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_small__gTGkU icon_hds-icon--size-s__2Lkik"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M11.9669 1.5C14.0265 1.5 16.0869 2.27765 17.6579 3.83367C20.8001 6.945 20.5878 11.7938 17.6579 15.1017L17.0345 15.8107C14.4235 18.8021 12.9688 20.7706 11.9669 22.7477C10.8937 20.6185 9.28527 18.4993 6.27797 15.1017C3.34814 11.7938 3.13508 6.945 6.27797 3.83367C7.84906 2.27765 9.90869 1.5 11.9669 1.5ZM11.999 3.5C10.3743 3.5 8.84685 4.12598 7.69956 5.26179C6.56645 6.38341 5.97897 7.81769 6.00058 9.41066C6.02151 10.984 6.64884 12.5453 7.76709 13.8087L8.42773 14.5606C10.0167 16.3842 11.1442 17.8 12.0004 19.0644C12.9748 17.6274 14.2916 16.0006 16.233 13.8087C17.3512 12.5453 17.9786 10.984 17.9995 9.41066C18.0198 7.81769 17.4323 6.38341 16.3005 5.26179C15.1519 4.12598 13.6251 3.5 11.999 3.5ZM12.0004 6C13.9338 6 15.5 7.56702 15.5 9.50039C15.5 11.433 13.9338 13 12.0004 13C10.067 13 8.5 11.433 8.5 9.50039C8.5 7.56702 10.067 6 12.0004 6ZM12.0003 8C11.1725 8 10.5 8.67319 10.5 9.50034C10.5 10.3268 11.1725 11 12.0003 11C12.8275 11 13.5 10.3268 13.5 9.50034C13.5 8.67319 12.8275 8 12.0003 8Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
        <h3
          class="text--bold text--h5"
        >
          Alueiden sijainti
        </h3>
      </div>
    </header>
    <div
      class="mapContainer"
    >
      <div
        class="mapContainer__inner"
        id="ol-map"
        style="height: 100%;"
      />
    </div>
  </div>
  <div
    class="css-zkgtxt"
  >
    <div
      aria-expanded="false"
      class="css-1ka79lz"
    >
      <button
        class="css-12ax0h3"
        type="button"
      >
        <div
          class="css-usczae"
        >
          <div>
            <h2
              class="heading-xs css-1eif5ff"
            >
              Hankealue 2
            </h2>
            <div
              class="css-aslqeu"
            >
              <p>
                15.5.2023
                –
                30.9.2023
              </p>
            </div>
          </div>
        </div>
        <svg
          aria-hidden="true"
          aria-label="angle-down"
          class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_small__gTGkU icon_hds-icon--size-s__2Lkik"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12 13.5L17 8.5L18.5 10L12 16.5L5.5 10L7 8.5L12 13.5Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
    <div
      style="display: none;"
    >
      <div
        class="css-r09jf8"
      >
        <div
          class="css-1evobf6"
        >
          <p
            class="text--bold text--body-m text--spacing-bottom-2-xs"
          >
            Työalue 1
             (
            159 m²
            )
          </p>
          <p>
            12.1.2023
            –
            12.11.2024
          </p>
        </div>
        <div
          class="css-1evobf6"
        >
          <p
            class="text--bold text--body-m text--spacing-bottom-2-xs"
          >
            Työalue 2
             (
            31 m²
            )
          </p>
          <p>
            12.1.2023
            –
            12.11.2024
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
